src/common/transformations/src/transformations/common_optimizations/shared_ops_optimization.cpp

```diff
@@ -0,0 +1,142 @@
+// Copyright (C) 2018-2023 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#include <openvino/core/validation_util.hpp>
+#include <openvino/op/concat.hpp>
+#include <openvino/op/gather_elements.hpp>
+#include <openvino/op/reshape.hpp>
+#include <openvino/op/slice.hpp>
+#include <openvino/op/tile.hpp>
+#include <openvino/op/util/sub_graph_base.hpp>
+#include <transformations/common_optimizations/shared_ops_optimization.hpp>
+
+#include "itt.hpp"
+
+using namespace std;
+using namespace ov;
+using namespace ov::op;
+
+bool shared_node_optimization(const shared_ptr<Model>& model,
+                              const unordered_map<Node::type_info_t, bool (*)(const Node*, const Node*)>& rules) {
+    bool rewritten = false;
+
+    for (const auto& op : model->get_ordered_ops()) {
```

Comment:
If I remember correctly, we decided to change the order of the traversal for this transformation (starting from Parameters and Constant and moving to Results and Sinks direction)

@t-jankowski could you remind us what was the reason for that?

Reply:
_could you remind us what was the reason for that?_
MSB transformation from PR #17694 removes (by merging or fusing) nodes from the graph and only after gathers next nodes top-down for further scanning. It's to avoid running multiple scanners on dangling node.
That's the case here indeed, but rather harmless - [if (target_inputs.size() <= 1) continue;](https://github.com/openvinotoolkit/openvino/pull/18622/files#diff-cff48c943726972afbe6e7ebeccc97205c793d1367845d2e3af3765327f0ada1R34-R35).

Reply:
My logic was designed with current topological order -- no need to alter it in any way.